aws_login.sh

```diff
@@ -15,30 +15,71 @@ function aws_login() {
+    if [ -z "$temporary_rackspace_token" ]; then
+      get_rackspace_token_and_tenant
+    fi
 
-    temporary_rackspace_token=$1
-    tennant_id=$2
+    aws_accounts=$(curl --location 'https://accounts.api.manage.rackspace.com/v0/awsAccounts' \
+      --silent \
+      --header "X-Auth-Token: $temporary_rackspace_token" \
+      --header "X-Tenant-Id: $rackspace_tennant_id" | jq -r '.awsAccounts[] | .awsAccountNumber + "_" + .name' | sed 's/\r//' | sort)
+
+    echo "$aws_accounts" > "$config_dir/aws_accounts.txt"
+  }
 
-    if [ -f "$config_dir/aws_accounts.txt" ]; then
-      cat "$config_dir/aws_accounts.txt"
+  function get_rackspace_username_and_api_key() {
+    kpscript_executable=$(command -v kpscript)
+
+    if [ -z "$KEEPASS_FILE" ] || [ -z "$kpscript_executable" ]; then
+      # no Keepass in place --> ask the user
+      echo "Did not found your Keepass file or KPScript executable. Please enter your Rackspace credentials."
+
+      read -r -p 'Rackspace username: ' rackspace_username
+      read -r -sp 'Rackspace API key: ' rackspace_api_key
+      
+      echo ""
     else
-      aws_accounts=$(curl --location 'https://accounts.api.manage.rackspace.com/v0/awsAccounts' \
-        --silent \
-        --header "X-Auth-Token: $temporary_rackspace_token" \
-        --header "X-Tenant-Id: $tennant_id" | jq -r '.awsAccounts[] | .awsAccountNumber + "_" + .name' | sed 's/\r//' | sort)
+      # get credentials from Keepass
+      echo "Reading credentials from Keepass: $KEEPASS_FILE. Entry Rackspace (username + api-key field)."
 
-      echo "$aws_accounts" > "$config_dir/aws_accounts.txt"
-      echo "$aws_accounts"
+      read -r -sp 'Keepass Password: ' keepass_password
+      echo ""
+
+      rackspace_username=$($kpscript_executable -c:GetEntryString "${KEEPASS_FILE}" -Field:UserName -ref-Title:"Rackspace" -FailIfNoEntry -pw:"$keepass_password" | head -n1 )
+      rackspace_api_key=$($kpscript_executable -c:GetEntryString "${KEEPASS_FILE}" -Field:api-key -ref-Title:"Rackspace" -FailIfNoEntry -pw:"$keepass_password" | head -n1 )
     fi
   }
 
-  temporary_rackspace_token=$(jq -r '.access.token.id' <<<"$rackspace_token_json")
-  tennant_id=$(jq -r '.access.token.tenant.id' <<<"$rackspace_token_json")
+  function get_rackspace_token_and_tenant() {
+    get_rackspace_username_and_api_key
+
+    rackspace_token_json=$(curl --location 'https://identity.api.rackspacecloud.com/v2.0/tokens' \
+      --header 'Content-Type: application/json' \
+      --silent \
+      --data "{
+            \"auth\": {
+                \"RAX-KSKEY:apiKeyCredentials\": {
+                    \"username\": \"$rackspace_username\",
+                    \"apiKey\": \"$rackspace_api_key\"
+                }
+            }
+        }")
+
+    temporary_rackspace_token=$(jq -r '.access.token.id' <<<"$rackspace_token_json")
+    rackspace_tennant_id=$(jq -r '.access.token.tenant.id' <<<"$rackspace_token_json")
+  }
+
+  if [ ! -f "$config_dir/aws_accounts.txt" ]; then
```

Comment:
In case the user fails during this step, the file `aws_accounts.txt` is created but remains empty. This check only checks for presence of the file thus the accounts are never fetched again. We should add a check to see if the file actually contains something.

Maybe this condition fits better:
`[ ! -f "$config_dir/aws_accounts.txt" ] || [ -z "$(cat "$config_dir"/aws_accounts.txt)" ]`